picard/browser/browser.py

```diff
@@ -25,115 +25,169 @@
+    log,
+)
 from picard.config import get_config
 from picard.util import mbid_validate
+from picard.util.thread import to_main
+
+
+SERVER_VERSION = '%s-%s/%s' % (PICARD_ORG_NAME, PICARD_APP_NAME, PICARD_VERSION_STR)
+RE_VALID_ORIGINS = re.compile(r'^(?:[^\.]+\.)*musicbrainz\.org$')
 
 
-def response(code):
-    if code == 200:
-        resp = '200 OK'
-    elif code == 400:
-        resp = '400 Bad Request'
-    else:
-        resp = '500 Internal Server Error'
-    return bytearray(
-        'HTTP/1.1 {}\r\n'
-        'Cache-Control: max-age=0\r\n'
-        '\r\n'
-        'Nothing to see here.\r\n'.format(resp), 'ascii')
+def _is_valid_origin(origin):
+    try:
+        url = urlparse(origin)
+    except ValueError:
+        return False
+    hostname = url.hostname
+    if not hostname:
+        return False
+    if RE_VALID_ORIGINS.match(hostname):
+        return True
+    config = get_config()
+    return config.setting['server_host'] == hostname
 
 
-class BrowserIntegration(QtNetwork.QTcpServer):
+class BrowserIntegration(QtCore.QObject):
 
-    """Simple HTTP server for web browser integration."""
+    listen_port_changed = QtCore.pyqtSignal(int)
 
     def __init__(self, parent=None):
         super().__init__(parent)
-        self.newConnection.connect(self._accept_connection)
-        self.port = 0
-        self.host_address = None
+        self.server = None
+
+    @property
+    def host_address(self):
+        if not self.server:
+            return ''
+        return self.server.server_address[0]
+
+    @property
+    def port(self):
+        if not self.server:
+            return 0
+        return self.server.server_address[1]
 
     def start(self):
-        if self.port:
+        if self.server:
             self.stop()
 
         config = get_config()
         if config.setting["browser_integration_localhost_only"]:
-            self.host_address = QtNetwork.QHostAddress(QtNetwork.QHostAddress.LocalHost)
+            host_address = '127.0.0.1'
         else:
-            self.host_address = QtNetwork.QHostAddress(QtNetwork.QHostAddress.Any)
+            host_address = '0.0.0.0'  # nosec
 
         for port in range(config.setting["browser_integration_port"], 65535):
-            if self.listen(self.host_address, port):
-                log.debug("Starting the browser integration (%s:%d)", self.host_address.toString(), port)
-                self.port = port
-                self.tagger.listen_port_changed.emit(self.port)
-                break
+            try:
+                self.server = HTTPServer((host_address, port), RequestHandler)
+            except OSError:
+                continue
+            log.info("Starting the browser integration (%s:%d)", host_address, port)
+            self.listen_port_changed.emit(port)
+            threading.Thread(target=self.server.serve_forever).start()
+            break
+        else:
+            log.error("Failed finding an available port for the browser integration.")
```

Comment:
Perhaps add tried range in this message, and while at it, store 65535 in a variable:

```python
min_port = config.setting["browser_integration_port"]
max_port = 65535
for port in range(min_port, max_port):
    ...
else:
    log.error("Failed finding an available port in range %d:%d for the browser integration.", min_port, max_port)
```

We don't need to check `min_port` <= `max_port`, right?

Reply:
No. Codewise it would just be an empty list and run into the else. Practically this is avoided as we restrict the values ` browser_integration_port` can be set to.